File: src/java/src/main/java/org/elastos/hive/vault/database/QueryOptions.java
package org.elastos.hive.vault.database;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.gson.annotations.SerializedName;

import java.util.Collections;
import java.util.List;

public class QueryOptions {
	private Integer skip;
	private Integer limit;
	private JsonNode projection;
	// ex: [('_id', -1)]
	private List<SortItem> sort;
	@SerializedName("allow_partial_results")
	private Boolean allowPartialResults;
	@SerializedName("return_key")
	private Boolean returnKey;
	@SerializedName("show_record_id")
	private Boolean showRecordId;
	@SerializedName("batch_size")
	private Integer batchSize;

	public QueryOptions setSkip(Integer skip) {
		this.skip = skip;
		return this;
	}

	public QueryOptions setLimit(Integer limit) {
		this.limit = limit;
		return this;
	}

	public QueryOptions setProjection(JsonNode projection) {
		this.projection = projection;
		return this;
	}

	public QueryOptions setSort(List<SortItem> sort) {
		this.sort = sort;
		return this;
	}

	public QueryOptions setSort(SortItem sort) {
		this.sort = Collections.singletonList(sort);
		return this;
	}

	public QueryOptions setAllowPartialResults(Boolean allowPartialResults) {
		this.allowPartialResults = allowPartialResults;
		return this;
	}

	public QueryOptions setReturnKey(Boolean returnKey) {
		this.returnKey = returnKey;
		return this;
	}

	public QueryOptions setShowRecordId(Boolean showRecordId) {
		this.showRecordId = showRecordId;
		return this;
	}

	public QueryOptions setBatchSize(Integer batchSize) {
		this.batchSize = batchSize;
		return this;
	}

	public Integer getSkip() {
		return skip;
	}

	public Integer getLimit() {
		return limit;
	}

	public JsonNode getProjection() {
		return projection;
	}

	public List<SortItem> getSort() {
		return sort;
	}

	public Boolean getAllowPartialResults() {
		return allowPartialResults;
	}

	public Boolean getReturnKey() {
		return returnKey;
	}

	public Boolean getShowRecordId() {
		return showRecordId;
	}

	public Integer getBatchSize() {
		return batchSize;
	}
}

File: src/service/database/queryoptions.ts
import {JSONObject} from "@elastosfoundation/did-js-sdk";
import {SortItem} from "./sortitem";
import {FindOptions} from "./findoptions";

/**
 * The options of the querying operation.
 */
export class QueryOptions extends FindOptions {
	private projection: JSONObject;
    private sort: [string, number][];
	private allow_partial_results: boolean;
	private return_key: boolean;
	private show_record_id: boolean;
	private batch_size: number;

    setProjection(projection: JSONObject) {
        this.projection = projection;
        return this;
    }

    setSort(items: SortItem[]) {
        this.sort = items ? items.map((t) => [t.getKey(), t.getOrder()]) : undefined;
        return this;
    }

    setAllowPartialResults(isAllow: boolean) {
        this.allow_partial_results = isAllow;
        return this;
    }

    setReturnKey(isReturn: boolean) {
        this.return_key = isReturn;
        return this;
    }

    setShowRecordId(isShow: boolean) {
        this.show_record_id = isShow;
        return this;
    }

    setBatchSize(size: number) {
        this.batch_size = size;
        return this;
    }

    getProjection(): JSONObject {
        return this.projection;
    }

    getSort(): [string, number][] {
        return this.sort;
    }

    isAllowPartialResults(): boolean {
        return this.allow_partial_results;
    }

    isReturnKey(): boolean {
        return this.return_key;
    }

    isShowRecordId(): boolean {
        return this.show_record_id;
    }

    getBatchSize(): number {
        return this.batch_size;
    }
}
